File: Algorithms/238. Product of Array Except Self.rs
impl Solution {
    pub fn product_except_self(nums: Vec<i32>) -> Vec<i32> {
        let length = nums.len();
        let mut res: Vec<i32> = vec![1; length];
        
        let mut product = 1;
        for i in 0..length {
            res[i] *= product;
            product *= nums[i];
        }
        
        product = 1;
        for i in (0..length).rev() {
            res[i] *= product;
            product *= nums[i];
        }
        
        res
    }
}

File: Algorithms/238. Product of Array Except Self.ts
function productExceptSelf(nums: number[]): number[] {
  const n = nums.length;
  const result: number[] = new Array(n);
  for (let i = 0, left = 1; i < n; i++) {
    result[i] = left;
    left *= nums[i];
  }
  for (let i = n - 1, right = 1; i >= 0; i--) {
    result[i] *= right;
    right *= nums[i];
  }
  return result;
}
